Content.Server/_Impstation/TraitRandomizer/TraitRandomizerSystem.cs

```diff
@@ -0,0 +1,127 @@
+    [Dependency] private readonly IServerPreferencesManager _prefs = default!;
+    [Dependency] private readonly EntityWhitelistSystem _whitelistSystem = default!;
+    [Dependency] private readonly SharedMindSystem _mind = default!;
+    [Dependency] private readonly SharedBodySystem _bodySystem = default!;
+    [Dependency] private readonly SharedHandsSystem _sharedHandsSystem = default!;
+    [Dependency] private readonly TagSystem _tagSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<TraitRandomizerComponent, MapInitEvent>(OnMapInit);
+    }
+
+    private void OnMapInit(Entity<TraitRandomizerComponent> ent, ref MapInitEvent args)
+    {
+        if (!_mind.TryGetMind(ent, out _, out var mindComponent) || mindComponent.Session == null)
+            return;
+
+        var allTraits = _prototypeManager.EnumeratePrototypes<TraitPrototype>().ToList();
+        List<TraitPrototype> traits = [];
+
+        // make a list of the traits we should be adding 
+        foreach (var trait in allTraits)
+        {
+            if (ent.Comp.Fonts && trait.Category == _prototypeManager.Index<TraitCategoryPrototype>("SignatureFonts"))
+                traits.Add(trait);
+
+            if (ent.Comp.Accents && trait.Category == _prototypeManager.Index<TraitCategoryPrototype>("SpeechTraits"))
+                traits.Add(trait);
+
+            if (ent.Comp.Quirks && trait.Category == _prototypeManager.Index<TraitCategoryPrototype>("Quirks"))
+                traits.Add(trait);
+
+            if (ent.Comp.Disabilities && trait.Category == _prototypeManager.Index<TraitCategoryPrototype>("Disabilities"))
+                traits.Add(trait);
+
+        }
+
+        var curProfile = (HumanoidCharacterProfile)_prefs.GetPreferences(mindComponent.Session.UserId).SelectedCharacter;
+
+        var curTraits = curProfile.TraitPreferences.ToList();
+
+        // remove currently applied traits from the list of traits we can roll from.
+        foreach (var traitProto in curTraits)
+        {
+            var trait = _prototypeManager.Index(traitProto);
+            traits.Remove(trait);
+        }
+
+        // how many traits are we gonna get?
+        var traitsToRoll = _random.Next(ent.Comp.MinTraits, ent.Comp.MaxTraits);
+        List<TraitPrototype> finalTraits = [];
+
+        // pick a trait, ensure we don't pick it again, and add it to the final traits list. do this that many times.
+        // note: currently this ignores points limits, because I think it's funnier that way. 
+        for (var i = 0; i < traitsToRoll; i++)
+        {
+            var thisTrait = _random.Pick(traits);
+            allTraits.Remove(thisTrait);
+            finalTraits.Add(thisTrait);
+        }
+
+        foreach (var traitId in finalTraits)
+        {
+            if (!_prototypeManager.TryIndex<TraitPrototype>(traitId, out var traitPrototype))
```

Comment:
gonna be a problem if upstream changes `TraitSystem`'s code - i could tell this was copypasted from somewhere